src/roofs.cpp

```diff
@@ -0,0 +1,291 @@
+#include "roofs.h"
+#include "map.h"
+#include "game.h"
+#include "mapdata.h"
+#include "line.h"
+#include <array>
+#include <algorithm>
+
+constexpr size_t padded_map_size = ( SEEX + 2 ) * MAPSIZE * ( SEEY + 2 ) * MAPSIZE;
+
+static const std::array < tripoint, ( size_t )roof_dir::num_dirs + 1 > dir_offsets = {{
+        tripoint( -1, 0, 0 ),
+        tripoint( 0, -1, 0 ),
+        tripoint( 1, 0, 0 ),
+        tripoint( 0, 1, 0 ),
+        tripoint( 0, 0, 0 )
+    }
+};
+
+roof_dir opposite_dir( roof_dir dir )
+{
+    switch( dir ) {
+        case roof_dir::west:
+            return roof_dir::east;
+        case roof_dir::north:
+            return roof_dir::south;
+        case roof_dir::east:
+            return roof_dir::west;
+        case roof_dir::south:
+            return roof_dir::north;
+        case roof_dir::num_dirs:
+            return roof_dir::num_dirs;
+    }
+    return roof_dir::num_dirs;
+}
+
+roof_dir convert_directions( direction dir )
+{
+    switch( dir ) {
+        case NORTH:
+            return roof_dir::north;
+        case EAST:
+            return roof_dir::east;
+        case SOUTH:
+            return roof_dir::south;
+        case WEST:
+            return roof_dir::west;
+        default:
+            break;
+    }
+    return roof_dir::num_dirs;
+}
+
+direction convert_directions( roof_dir dir )
+{
+    switch( dir ) {
+        case roof_dir::north:
+            return NORTH;
+        case roof_dir::east:
+            return EAST;
+        case roof_dir::south:
+            return SOUTH;
+        case roof_dir::west:
+            return WEST;
+        default:
+            break;
+    }
+    return CENTER;
+}
+
+const tripoint &direction_to_offset( roof_dir dir )
+{
+    return dir_offsets[ ( size_t )dir ];
+}
+
+struct cached_tile {
```

Comment:
Does this need a cache?  How long does it take to decide whether a wall or floor section can be safely removed?
Also I guess how long are redraws of roof debug view, though I'd rather tolerate that being kind of slow than add otherwise unnecessary caching.